mallocMC: Definitions, Component and Library Support

slizzered opened the pull request:
- REQUIRED and QUIET work as intended
- halloc is available as a COMPONENT (beta)
- see #3 for the full writeup

To circumvent the path problem, it could be used in a `CMakeLists.txt` like this:

``` cmake
find_package(mallocMC 2.1.0 QUIET)

if(NOT mallocMC_FOUND)
  message(STATUS "Using mallocMC from thirdParty/ directory")
  set(MALLOCMC_ROOT "../../thirdParty/mallocMC")
  find_package(mallocMC 2.1.0 REQUIRED)
endif(NOT mallocMC_FOUND)

include_directories(SYSTEM ${mallocMC_INCLUDE_DIRS})
add_definitions(${mallocMC_DEFINITIONS})
# append ${mallocMC_LIBRARIES} to your target libraries, e.g.,
# set(LIBS ${LIBS} ${mallocMC_LIBRARIES})
#   or
# list(APPEND LIBS ${mallocMC_LIBRARIES})
```

slizzered committed:
Restructure module FindmallocMC.cmake

 - REQUIRED and QUIET work as intended
 - halloc is available as a COMPONENT (beta)
 - see #3 for the full writeup

slizzered committed:
removed hardcoded path

ax3l commented:
that is a great update and the `set(CUDA_SEPARABLE_COMPILATION ON PARENT_SCOPE)` is a wonderful solution. I will have a look on the details of the change set soon :)

ax3l commented:
Shouldn't it be `CUDA_SEPARABLE_COMPILATION` and is it now required to use `find_package` in `mallocMC` before checking CUDA in a project to make it work?

slizzered commented:
I don't quite understand the question about `CUDA_SEPARABLE_COMPILATION`...

About the `find_package` calls:
Previously, we required per convention/comment, that the call `find_package(mallocMC)` was only allowed _after_ CUDA and Boost were already found by the calling `CMakeLists.txt`. I don't see any real downside in executing `find_package` inside the module as well, since the calls appear to be cumulative.
That way, the call `find_package(mallocMC)` can be placed wherever we want in the `CMakeLists.txt`. If more modules of Boost or a higher CUDA version are required, the `CMakeLists.txt` can still do this through additional calls to `find_package(...)`.
If I overlooked a case where the new behavior causes problems, we can go back to the old style and use comments.

BenjaminW3 commented on FindmallocMC.cmake:
```diff
@@ -41,66 +45,48 @@
-  DOC "mallocMC ROOT location"
-)
+    NAMES include/mallocMC/mallocMC.hpp
+    PATHS ENV MALLOCMC_ROOT
```
Could you add a possibility to provide the MallocMC root directory through a CMake variable in addition to the environment variable?
This would simplify the usage example given in your description and would make it more consistent to other Find scripts e.g. FindBoost with its BOOST_ROOT.
Furthermore this would allow to compile a project using MallocMC with different MallocMC variants only by changing the cmake command line.

ax3l replied:
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

BOOST_ROOT can be env. _or_ cmake? Did not know that, interesting. most modules only allow one of those.

Besides that _ROOT is a bit deprecated and for env. CMAKE_PREFIX_PATH is the more modern way.
-----BEGIN PGP SIGNATURE-----
Version: APG v1.1.1

iQJCBAEBCgAsBQJViBbDJRxBeGVsIEh1ZWJsIDxheGVsLmh1ZWJsQHBsYXNtYS5u
aW5qYT4ACgkQRy0sNk4h2UtPxQ//ava4uCKhDaU/pEnJGuwN4woSw0rF1tt+AlQd
Mlgf4scpFTuNGNX4mtt8pJM1MPyIErSKSil1SufAOjiLyU8UHLRybiSpvgKWoq8K
FObrKyH07B/JPH6BCH26YQ9+pqjhkMHEoChzzER+aAMXXhBMdpiOir14tLZ+ZzOk
EkGJkF/Vv4Hin07jVY3LoUqwx0hhmX7FbH7ii/uJHLWz2xUDcumzc3TET/bXEtsT
5E+/yiwa+yTY9mrpNnnKQ6GETVSgvCk4wsS/4jCOhfSlpncFSoKx5knUk1pZqxYm
FMpUEbcNWK3JT4V9Mw5uogfVK6/+4oL6OyYEPYAnHNueAMwwGDRSu5wyG1VHvUbP
mcWVEaUVVt4MpE7r+nC8gv+smE3lnTLyronmQGOhRP0SmN+kP6enbITJl0h2MvJ0
Y1+MBUOodMSSUj7fI1ZUVInYs8yIG37krRp4DR5Ax+v8ubuHcXL8K6s23O/Nh+3K
eqenXq3jT1R8JI8kn8JoLFnFbp/ekHxOGe7hxU4s2q6EiRyAWEmFmoR5NF6kdQ47
6fyruiugKQrVhao7uvemmUXQLr/vKZEGwNI703o5gB6Eks8+WgBweNEcfPEvSKZY
WrHYJUg/BoBXFeQ2jaVEhlfxMh3hNY4N77Ts4E9SZwvaVzRZlMnzrK8OKI0DyGor
zWNVW74=
=2ioU
-----END PGP SIGNATURE-----

slizzered committed:
Also use CMake variable MALLOCMC_ROOT

 - The module `FindmallocMC.cmake` now searches first in the CMake
   variable `MALLOCMC_ROOT` and, if not successful, in the environment
   variable `MALLOCMC_ROOT`.

slizzered commented:
@BenjaminW3 : good idea, see updated code. In addition to the environment variable, you can now also use a simple

``` cmake
set(MALLOCMC_ROOT "../../thirdParty/mallocMC")
```

This variable takes precedence over the environment variable of the same name (`$ENV{MALLOCMC_ROOT}`)

ax3l commented:
> This variable takes precedence over the environment variable of the same name ($ENV{MALLOCMC_ROOT})

Can you please add this behavior to the documentation of the `_ROOT` variable(e) in line 13-14 of this module? It needs documentation.

ax3l commented:
I added the lines

``` cmake
add_definitions(${mallocMC_DEFINITIONS})
# append ${mallocMC_LIBRARIES} to your target libraries, e.g.,
# set(LIBS ${LIBS} ${mallocMC_LIBRARIES})
#   or
# list(APPEND LIBS ${mallocMC_LIBRARIES})
```

to your pull request description. @slizzered Is that correct?

slizzered committed:
Added documentation about MALLOCMC_ROOT variable

slizzered commented:
@ax3l yes, the lines are correct. 
However, the module currently does not provide any definitions, therefore the line `add_definitions(${mallocMC_DEFINITIONS})` could also be commented out without negative effects (simliar to `${mallocMC_LIBRARIES}`)

ax3l commented:
I know that it has no effect, but the idea is to provide it already so
user-code (CMakeLists.txt) does not need changes when they get an effect
in the future :)

ax3l commented:
Doe I have to call `find_package(mallocMC)` before a general `find_package(CUDA)` call or doesn't it matter?

I think I will also continue to use

``` cmake
set(MALLOCMC_ROOT "${CMAKE_CURRENT_SOURCE_DIR}/../../thirdParty/mallocMC")
```

since I remember problems with a bare `../../` with out-of-source builds.

ax3l commented on FindmallocMC.cmake:
```diff
@@ -41,66 +47,48 @@
+    # mallocMC variables ########################################################
     #
     list(APPEND mallocMC_INCLUDE_DIRS ${mallocMC_ROOT_DIR}/include)
+    list(APPEND mallocMC_DEFINITIONS " ")
```
regression: these aren't _emtpy lists_ but lists with an element `" "`.
-> else the compiler will try to add an empty `-l [space]` somewhere in the linker string when added to target_libraries
